content/blog/2022-07-05-serving-models-with-mlem.md

```diff
@@ -0,0 +1,283 @@
+
+from mlem.api import save
+import numpy as np
+from sklearn.ensemble import RandomForestClassifier
+...
+```
+
+Then you can add the `save` function to the end of the training script.
+
+```python
+# train.py
+...
+clf = RandomForestClassifier(
+    n_estimators=n_est, min_samples_split=min_split, n_jobs=2, random_state=seed
+)
+
+clf.fit(x, labels)
+
+save(
+    clf,
+    "clf",
+    sample_data=x,
+    description="Random Forest Classifier",
+)
+```
+
+<admon type="tip">
+
+You don't have to do these steps as we already have a model available, but if
+you want to see the training and evaluation steps in action, you can run a few
+experiments with:
+
+```dvc
+$ dvc exp run
+```
+
+This will execute the DVC pipeline and you can check out what is happening in
+that pipeline by looking at the `dvc.yaml` file in the project.
+
+You can also see where we load the model into the `src/evaluate.py` script. To
+do that, you'll need to add the following import.
+
+```python
+# evaluate.py
+
+import json
+import math
+import os
+import sys
+
+import pickle5 as pickle
+import sklearn.metrics as metrics
+import numpy as np
+from mlem.api import apply
+...
+```
+
+Now we can use the [`apply` function](https://mlem.ai/doc/api-reference/apply)
+to make predictions with the model.
+
+```python
+# evaluate.py
+...
+x = matrix.iloc[:,1:11].values
+
+cleaned_x = np.where(np.isnan(x), 0, x)
+labels_pred = apply(model_file, cleaned_x, method="predict")
+
+predictions_by_class = apply(model_file, cleaned_x, method="predict_proba")
+predictions = predictions_by_class[:, 1]
+...
+```
+
+The `predict` and `predict_proba` are methods available from the model and they
+are used to get new predicted values and their probabilities for evaluation.
+This, along with everything else in the script, is how we get the metrics for
+each experiment.
+
+</admon>
+
+After you run an experiment, there will be two new files in your repo: `clf` and
+`clf.mlem`. Make sure you add the `clf.mlem` file to your Git history with the
+following command:
+
+```dvc
+$ git add clf.mlem
+```
+
+This is so that the metadata is in your repo and ready to use with other MLEM
+commands. Now we can finally take the model file and ship it to production.
+
+## Serve the model to production
+
+There are a couple of ways you can do this with MLEM:
+
+- Serve the model with [FastAPI](https://fastapi.tiangolo.com/)
+- Create a Python package
+
+_Note:_ There is an experimental option to
+[deploy the model directly to Heroku](https://mlem.ai/doc/get-started/deploying)
+although this functionality is experimental and may have breaking changes.
+
+### Serve with FastAPI
+
+If you don't have an API to work with and don't need a Python package, you can
+serve your model quickly using FastAPI with this command.
+
+```dvc
+$ mlem serve clf fastapi
+```
+
+This will run a local server and spin up an API for you so you can quickly test
+out your model without needing a development team to work on the API initially.
```

Comment:
💅🏼 (API could be so many different things...)
```suggestion
This will run a local server and spin up a web API for you so you can quickly test
out your model without needing a development team to work on the API initially.
```